core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala

```diff
@@ -14,233 +14,532 @@
+            memberId = memberId,
+            memberEpoch = memberEpoch,
+            partitions = null,
+            requireStable = requireStable,
+            version = version.toShort
+          )
+        )
+
+        if (useNewProtocol && version >= 9) {
+          assertEquals(
+            new OffsetFetchResponseData.OffsetFetchResponseGroup()
+              .setGroupId("grp")
+              .setErrorCode(Errors.UNKNOWN_MEMBER_ID.code),
+            fetchOffsets(
+              groupId = "grp",
+              memberId = "",
+              memberEpoch = memberEpoch,
+              partitions = null,
+              requireStable = requireStable,
+              version = version.toShort
+            )
+          )
+
+          assertEquals(
+            new OffsetFetchResponseData.OffsetFetchResponseGroup()
+              .setGroupId("grp")
+              .setErrorCode(Errors.STALE_MEMBER_EPOCH.code),
+            fetchOffsets(
+              groupId = "grp",
+              memberId = memberId,
+              memberEpoch = memberEpoch + 1,
+              partitions = null,
+              requireStable = requireStable,
+              version = version.toShort
+            )
+          )
         }
-        verifySingleGroupResponse(version.asInstanceOf[Short],
-          response.error().code(), partitionData.errorCode(), topicData.name(),
-          partitionData.partitionIndex(), partitionData.committedOffset(),
-          partitionData.committedLeaderEpoch(), partitionData.metadata())
-      } else {
-        val request = new OffsetFetchRequest.Builder(
-          Collections.singletonMap(groupId, null),
-          false,
-          false
-        ).build(version.toShort)
-
-        val response = connectAndReceive[OffsetFetchResponse](request)
-        assertEquals(Errors.NONE, response.groupLevelError(groupId))
-        val groupData = response.data().groups().get(0)
-        val topicData = groupData.topics().get(0)
-        val partitionData = topicData.partitions().get(0)
-        verifySingleGroupResponse(version.asInstanceOf[Short],
-          groupData.errorCode(), partitionData.errorCode(), topicData.name(),
-          partitionData.partitionIndex(), partitionData.committedOffset(),
-          partitionData.committedLeaderEpoch(), partitionData.metadata())
+
       }
     }
   }
 
-  @Test
-  def testOffsetFetchRequestWithMultipleGroups(): Unit = {
-    createTopic(topics(0))
-    createTopic(topics(1), numPartitions = 2)
-    createTopic(topics(2), numPartitions = 3)
-
-    // create 5 consumers to commit offsets so we can fetch them later
-    val partitionMap = groupToPartitionMap.asScala.map(e => (e._1, Option(e._2).getOrElse(allTopicsList)))
-    groups.foreach { groupId =>
-      consumerConfig.setProperty(ConsumerConfig.GROUP_ID_CONFIG, groupId)
-      commitOffsets(partitionMap(groupId))
+  private def testMultipleGroupsOffsetFetch(useNewProtocol: Boolean, requireStable: Boolean): Unit = {
```

Comment:
Another "my understanding" question. 😅 Do only admin clients request multiple group offsets at once? I probably should have realized this earlier, but I see we don't pass member id or epoch in this version of the request.

Reply:
Yeah, that's correct.